`buildPythonPackage` , `buildPythonApplication`: take `installCheck`-related attributes directly

ShamrockLee opened the pull request:
## Description of changes

This draft PR enables `buildPythonPackage` and `buildPythonApplicaiton` to directly take the `installCheck`- related attribute instead of specifying it using `check`- associated arguments.

Following the community consensus to merge `lib.extendMkDerivation`, there are ongoing efforts to make `buildPythonPackage` and `buildPythonApplication` take  fixed-point arguments (`(finalAttrs: { })`) and supports the unified overrider `<pkg>.overrideAttrs`.

As the finalAttrs are provided by stdenv.mkDerivation, it contains `doInstallCheck` and `installCheckInputs` instead of `doCheck` and `checkInput`. Should `buildPythonPackage` continues to take only the `checkPhase`-related arguments and map them to the `installCheckPhase`-related ones, it creates inconsistency between the specification and reference of the same attribute.

For example, with fixed-point arguments and `overrideAttrs`, there will be package definitions like:

```nix
{
  buildPythonPackage
}:
buildPythonPackage (finalAttrs: {
  doCheck = false;
  checkInputs = lib.optionals finalAttrs.doInstallCheck [
    # installCheckInputs elements
  ];
  passthru = {
    tests = {
      python-check = finalAttrs.finalPackage.overrideAttrs {
        doInstallCheck = true;
      };
    };
  };
});
```

This PR will also fulfill the to-do stated in [the `mk-python-derivation` comment](https://github.com/NixOS/nixpkgs/blob/2d49e5c54ef320820066a3269321dc21957cdb99/pkgs/development/interpreters/python/mk-python-derivation.nix#L340) to deprecate the use of `checkPhase` to specify `installCheckPhase`, and use `installCheckPhase` directly.

This PR depends on
- #379637

which is a staging change.

After the depending PR merged and this PR rebased, it should create only a few rebuilds.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ShamrockLee committed:
doc python.section.md: specify installCheck-related attributes directly

Change checkPhase to installCehckPhase, doCheck to doInstallCheck, ...,
to reflect the fact that the specification goes to the
installCheck-related attributes and that buildPython* build helpers now
takes the installCheck-related attributes directly in addition to
indirectly specify them thourgh the check-related arguments.

Co-authored-by: Alexis Hildebrandt <afh@surryhill.net>

ShamrockLee committed:
erosmb: fix do-test specification

In the previous implementation of buildPythonApplication, the
doInstallCheck attribute was always specified by the doCheck argument.
as `doCheck = false` was specified by this package, there's no effect to
specify `doInstallCheck = true`, and the `installCheckPhase` specified
doesn't get executed during build.

After buildPythonApplication takes `doInstallCheck` directly,
`doInstallCheck = true` starts to take effect. The `doCheck` argument,
as an alias to `doInstallCheck`, lost its effect when `doInstallCheck`
is presented.

The setuphookCheckPhase are still supressed due to the fact that
installCheckPhase is specified non-empty.

ShamrockLee committed:
Add release note entry about buildPython* taking installCheck-relatd attributes

Co-authored-by: Martin Weinelt <mweinelt@users.noreply.github.com>

afh commented on doc/languages-frameworks/python.section.md:
```diff
@@ -144,8 +144,10 @@ The `buildPythonPackage` mainly does four things:
+By default tests are run because [`doInstallCheck = true`](#var-stdenv-doInstallCheck). Test dependencies, like
+e.g. the test runner, should be added to [`nativeInstallCheckInputs`](#var-stdenv-nativeInstallCheckInputs).
+
+Python packages and applicationst doesn't have [`checkPhase`](#ssec-check-phase). `doCheck`, `checkPhase`, `checkInputs`, and `nativeCheckInputs` are aliases to the `installCheck`-related attributes for historical reason, but using them instead of the `installCheck`-related variables are obsolete and discouraged.
```
Minor spelling and grammar suggestion:
```suggestion
Python packages and applications don't have [`checkPhase`](#ssec-check-phase). `doCheck`, `checkPhase`, `checkInputs`, and `nativeCheckInputs` are aliases to the `installCheck`-related attributes for historical reason, but using them instead of the `installCheck`-related variables are obsolete and discouraged.
```

afh commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1936,9 +1938,9 @@ Both are also exported in `nix-shell`.
-The best practice today is to pass a test hook (e.g. pytestCheckHook, unittestCheckHook) into nativeCheckInputs.
-This will reconfigure the checkPhase to make use of that particular test framework.
-Occasionally packages don't make use of a common test framework, which may then require a custom checkPhase.
+The best practice today is to pass a test hook (e.g. pytestCheckHook, unittestCheckHook) into nativeInstallCheckInputs.
+This will reconfigure the installCheckPhase to make use of that particular test framework.
+Occasionally packages don't make use of a common test framework, which may then require a custom installCheckPhase.
```
Minor formatting and linking suggestions:
```suggestion
The best practice today is to pass a test hook (e.g. `pytestCheckHook`, `unittestCheckHook`) into [`nativeInstallCheckInputs`](#var-stdenv-nativeInstallCheckInputs).
This will reconfigure the `installCheckPhase` to make use of that particular test framework.
Occasionally packages don't make use of a common test framework, which may then require a custom `installCheckPhase`.
```

afh reviewed (commented):
Thanks for continuing the work on this, @ShamrockLee, please find below some minor suggestions…

mweinelt commented:
I believe to understand the distinction between `checkPhase` and `installCheckPhase` in the meaning of the stdenv chapter in the manual. What I don't understand is why Python adopted one over the other, and what the motivation and benefit of rolling that decision back is.

ShamrockLee commented:
@mweinelt The issue is that we are
- specifying the `installCherkPhase` attributes using the `checkPhase` arguments of `buildPython*` and
- pretending that we are using `checkPhase` in the corresponding section of Nixpkgs Manual.

It is not about the choice between `chcekPhase` and `installCheckPhase`, but to inform the choice to users of the bud helpers.

mweinelt commented:
Why then not move us into `checkPhase` if it is all the same? I don't see the value in adopting another lingo when a whole bunch of packages have already adopted it.

ShamrockLee commented:
That's indeed another way to go.

The decision to implement `doCheck` with  `doInstallCheck` dates back to 2016, in commit 3e05cce97ca3ea9eee42a17c8bc0b89345b5d0b6 made by @FRidh in #18143. The comment mentioning the long-term goal to move to `installCheckPhase` was added in 2019 in commit f7e28bf5d8181926e600a222cb70180519d09726.

natsukium commented:
> The issue is that we are
>
> - specifying the installCherkPhase attributes using the checkPhase arguments of buildPython* and
> - pretending that we are using checkPhase in the corresponding section of the Nixpkgs Manual.
>
> It is not about choosing between chcekPhase and installCheckPhase but about informing users of the build helpers about this choice.

The manual notes that checkPhase is used instead of installCheckPhase.
However, I do not know why we adopted checkPhase and not installCheckPhase as it was.
Perhaps because installCheckPhase was rarely used and unfamiliar to most people?

> The [checkPhase](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#ssec-check-phase) for python maps to the installCheckPhase on a normal derivation. This is due to many python packages not behaving well to the pre-installed version of the package. Version info, and natively compiled extensions generally only exist in the install directory, and thus can cause issues when a test suite asserts on that behavior.

https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#testing-python-packages-testing-python-packages

FRidh commented:
> I get that `installCheckPhase` runs post-install and is more useful for the reasons mentioned.
> 
> The frustrating thing here is, that we simplified the lingo to `build-system` and `dependencies`, only to move from `checkInputs` to `nativeCheckInputs`, and now to `nativeInstallCheckInputs`. I get that all of these bits and pieces have meaning, but we'll be dealing with the resulting churn for the next two releases at least.
> 
> And then there is the matter of [nix-community/nix-init#419](https://github.com/nix-community/nix-init/pull/419) 🫠

Build system and dependencies are essentially another layer on top. The underlying stdenv atttributes will always be needed for all other cases.

I think motivation at the time was consistency. Probably it was not given much thought.

mweinelt commented on nixos/doc/manual/release-notes/rl-2411.section.md:
```diff
@@ -193,6 +193,10 @@
 
 - The `stackclashprotection` hardening flag has been added, though disabled by default.
 
+- `buildPythonPackage` and `buildPythonApplication` now take `doCheck`-related attributes (`doInstallCheck`, `installCheckPhase`, `installCheckInputs`, `nativeInstallCheckInputs`) directly, which precedes the old approach of specifying them via the `check`-related arguments (`doCheck`, `checkPhase`, `checkInputs`, `nativeCheckInputs`).
+
+    Existing Python packages and applications will stay unchanged as long as they don't specify `doCheck` and `doInstallCheck` differently. Such specification is irrelevent both before and after this change, but the effect would be different.
```
```suggestion
- The `buildPythonPackage` and `buildPythonApplication` functions now consume the `installCheckPhase` attributes (`doInstallCheck`, `installCheckPhase`, `installCheckInputs`, `nativeInstallCheckInputs`) directly. `checkPhase` attributes (`doCheck`, `checkPhase`, `checkInputs`, `nativeCheckInputs`) have been aliases to their `installCheckPhase` counterparts since at least 2016 and their usage is now deprecated.
```

mweinelt commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -312,15 +312,15 @@ let
-    ] ++ nativeCheckInputs;
-    installCheckInputs = checkInputs;
+    ] ++ nativeInstallCheckInputs;
+    inherit installCheckInputs;
```
```suggestion

    inherit installCheckInputs;
```

mweinelt commented on pkgs/development/interpreters/python/python2/mk-python-derivation.nix:
```diff
@@ -209,15 +209,15 @@ let
-    ] ++ nativeCheckInputs;
-    installCheckInputs = checkInputs;
+    ] ++ nativeInstallCheckInputs;
+    inherit installCheckInputs;
```
```suggestion

    inherit installCheckInputs;
```

natsukium commented:
After some thought, I'm personally against the migration to `installCheckPhase` for now.
It has been over six months since we started the switch to `build-system/dependencies`, and this migration is painful.

Indeed, this change is correct and should be done to ensure consistency with the phase definition.
However, it will be more difficult to encourage contributors to migrate just for that reason, even if it is backwards compatible.

Regarding `installCheckPhase`, I feel that many of its functions have delegated `passthru.tests`, such as `tester.testVersion`.
It is no exception in python, where https://github.com/NixOS/nixpkgs/pull/272177 tries to separate build and test.
I don't know how it will finally be implemented, but it seems that it would be less burdensome if this PR change were adopted at the same time as the separation.

ShamrockLee commented:
> After some thought, I'm personally against the migration to `installCheckPhase` for now.
> It has been over six months since we started the switch to `build-system/dependencies`, and this migration is painful.
>
> Indeed, this change is correct and should be done to ensure consistency with the phase definition.
> However, it will be more difficult to encourage contributors to migrate just for that reason, even if it is backwards compatible.

Thank you for sharing your precious experiences about attribute renaming.

Your right. The change wouldn't be helpful to the community if the pain caused by the mass renaming outweighs the benefits of the name correctness.

> Regarding `installCheckPhase`, I feel that many of its functions have delegated `passthru.tests`, such as `tester.testVersion`.
> It is no exception in python, where #272177 tries to separate build and test.
> I don't know how it will finally be implemented, but it would be less burdensome if this PR change were adopted at the same time as the separation.

Delegating checks to `passthru.test` (as package tests) sounds like the right way to go. Let's push that one first.

If this is not the change we want to make now, should we close this pull request or mark it as a draft?

ShamrockLee committed:
python3Packages.buildPythonPackage: redirect pre/postCheck to pre/postInstallCheck

Goal: Unify the <pkg>.overrideAttrs interface for Python packages.

Specify preInstallCheck/postInstallCheck as preCheck/postCheck
if the latter is not specified.

Use preInstallCheck/postInstallCheck in buildPython* setup hooks.

When the checkPhase argument is specified,
replace `runHook pre/postCheck` with `runHook pre/postInstallCheck`
before assigning it to `installCheckPhase`.

ShamrockLee committed:
python3Packages.pytestCheckHook: assign pytestCheckPhase to installCheckPhase

Simplify implementation and ensure that
pre/postInstallCheck only run once.

ShamrockLee committed:
python3Packages.unittestCheckHook: assign unittestCheckPhase to installCheckPhase

Simplify implementation and ensure that
pre/postInstallCheck only run once.

ShamrockLee committed:
buildPythonPackage: take installCheck-related arguments directly

SandaruKasa commented:
Is it still/already relevant?

ShamrockLee commented:
This PR is surpassed by PR #469804, closing.

PR #469804 tells `makeOverridablePythonAttrs` (i.e., `<pkg>.overridePythonAttrs`) to handle both `checkPhase` and `installCheckPhase`, while passing only `installCheckPhase` to `stdenv.mkDerivation`.
Such approach is much more flexible and backward-compatible while ensuring the attribute correctness of `finalAttrs`.

I'll move the documentation change from this PR to PR #469804 or subsequent PRs.

- #469804

@SandaruKasa
